usr/lib/python3/dist-packages/stprint/tests/stprint.py

```diff
@@ -0,0 +1,457 @@
+            ("\x1b[;0;1;2;m", "\x1b[;0;1;2;m"),
+            ("\x1b[000001;000000000002;m", "\x1b[000001;000000000002;m"),
+            (
+                "\x1b[;0038;05;0001;000001;000038;005;00002;00038;05;0000003m",
+                "\x1b[;0038;05;0001;000001;000038;005;00002;00038;05;0000003m",
+            ),
+        ]
+        self.run_stprint_cases(cases, sgr=2**24)
+
+    def test_stprint_sgr_no_color(self) -> None:
+        """
+        Test without color
+        """
+        cases = [
+            ("\x1b[m", "_[m"),
+            ("\x1b[31m", "_[31m"),
+        ]
+        for sgr in (-1, -256, 0, 7):
+            self.run_stprint_cases(cases, sgr=sgr)
+
+    def test_stprint_sgr_three_bit(self) -> None:
+        """
+        Test with SGR 3-bit.
+        """
+        cases = [
+            ("\x1b[31m", "\x1b[31m"),
+            ("\x1b[91m", "_[91m"),
+            ("\x1b[4m", "_[4m"),
+            ("\x1b[38;5;1m", "_[38;5;1m"),
+            ("\x1b[38;2;255;0;1m", "_[38;2;255;0;1m"),
+        ]
+        self.run_stprint_cases(cases, sgr=2**3)
+
+    def test_stprint_sgr_four_bit(self) -> None:
+        """
+        Test with SGR 4-bit.
+        """
+        cases = [
+            ("\x1b[m", "\x1b[m"),
+            ("\x1b[31m", "\x1b[31m"),
+            ("\x1b[91m", "\x1b[91m"),
+            ("\x1b[4m", "_[4m"),
+            ("\x1b[38;5;1m", "_[38;5;1m"),
+            ("\x1b[38;2;255;0;1m", "_[38;2;255;0;1m"),
+        ]
+        self.run_stprint_cases(cases, sgr=2**4)
+
+    def test_stprint_sgr_eight_bit(self) -> None:
+        """
+        Test with SGR 88 colors and 8-bit.
+        """
+        cases = [
+            ("\x1b[m", "\x1b[m"),
+            ("\x1b[31m", "\x1b[31m"),
+            ("\x1b[91m", "\x1b[91m"),
+            ("\x1b[4m", "\x1b[4m"),
+            ("\x1b[38;5;1m", "\x1b[38;5;1m"),
+            ("\x1b[38;2;255;0;1m", "_[38;2;255;0;1m"),
+        ]
+        for sgr in (88, 2**8):
+            self.run_stprint_cases(cases, sgr=sgr)
+
+    def test_stprint_sgr_twenty_four_bit(self) -> None:
+        """
+        Test with SGR 24-bit.
+        """
+        cases = [
+            ("\x1b[m", "\x1b[m"),
+            ("\x1b[31m", "\x1b[31m"),
+            ("\x1b[91m", "\x1b[91m"),
+            ("\x1b[4m", "\x1b[4m"),
+            ("\x1b[38;5;1m", "\x1b[38;5;1m"),
+            ("\x1b[38;2;255;0;1m", "\x1b[38;2;255;0;1m"),
+        ]
+        self.run_stprint_cases(cases, sgr=2**24)
+
+    def test_stprint_no_extra_sgr(self) -> None:
+        """
+        Test disabling extra SGR.
+        """
+        cases = [
+            ("\x1b[31m", "\x1b[31m"),
+            ("\x1b[38;5;1m", "_[38;5;1m"),
+            ("\x1b[1;38;5;1;0m", "_[1;38;5;1;0m"),
+            ("\x1b[38;2;255;0;0m", "_[38;2;255;0;0m"),
+            ("\x1b[2;38;2;255;0;0;1m", "_[2;38;2;255;0;0;1m"),
+        ]
+        self.run_stprint_cases(
+            cases, sgr=2**24, exclude_sgr=["0*[3-4]8;+0*(2|5);+.*"]
+        )
```

Comment:
Matching more than one semicolon in a row after the 38/48 confuses me since a 38/48 code with more than one semicolon after it would be invalid anyway and thus rejected by stprint. Maybe we can remove the `+` from the `;+` in the middle and add a test that verifies the invalid code is rejected, something like this perhaps?

```suggestion
    def test_stprint_no_extra_sgr(self) -> None:
        """
        Test disabling extra SGR.
        """
        cases = [
            ("\x1b[31m", "\x1b[31m"),
            ("\x1b[38;5;1m", "_[38;5;1m"),
            ("\x1b[38;;5;1m", "_[38;;5;1m"),
            ("\x1b[1;38;5;1;0m", "_[1;38;5;1;0m"),
            ("\x1b[38;2;255;0;0m", "_[38;2;255;0;0m"),
            ("\x1b[2;38;2;255;0;0;1m", "_[2;38;2;255;0;0;1m"),
        ]
        self.run_stprint_cases(
            cases, sgr=2**24, exclude_sgr=["0*[3-4]8;0*(2|5);+.*"]
        )
```

Reply:
Fixed but added to another test.